### PR TITLE
add opensearch kubernetes tests

### PR DESCRIPTION
**Pull request by rawlingsj**
I've been trying to test this locally but the pods can't pull from registry 
```
Failed to pull image "k3d-k3d.localhost:5005/opensearch:unused@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d": failed to pull and unpack image "k3d-k3d.localhost:5005/opensearch@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d": failed to resolve reference "k3d-k3d.localhost:5005/opensearch@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d": failed to do request: Head "https://k3d-k3d.localhost:5005/v2/opensearch/manifests/sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d": dial tcp 127.0.0.1:5005: connect: connection refused
```

I can pull from my local machine

```
docker pull k3d-k3d.localhost:5005/opensearch:unused@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d
k3d-k3d.localhost:5005/opensearch@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d: Pulling from opensearch
Digest: sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d
Status: Image is up to date for k3d-k3d.localhost:5005/opensearch@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d
k3d-k3d.localhost:5005/opensearch:unused@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d

What's Next?
  View a summary of image vulnerabilities and recommendations → docker scout quickview k3d-k3d.localhost:5005/opensearch:unused@sha256:64df8f79688c754aa7001ea5cac5313f78326a559588ef18acd5c182e4119a9d
```

To build image and run tests
First 
```
make k3d
```
Then
```
IMAGETEST_SKIP_TEARDOWN=true TF_VAR_target_repository="k3d-k3d.localhost:5005" terraform apply -target=module.opensearch -auto-approve
```
I now grab the kubectl container ID from `docker ps`

Exec
```
docker exec -ti $PID bash
```

Then 
```
kubectl get pods -A
```
When pod starts failing 
```
kubectl describe pod  opensearch-cluster-master-0 -n opensearch
```

### CODE REVIEW

File: images/opensearch/config/template.apko.yaml [hunk 1]
```diff
@@ -20,9 +20,36 @@ paths:
...
+    uid: 65532
+    gid: 65532
+    permissions: 0o755
+    recursive: true
+  - path: /usr/share/opensearch/logs
+    type: directory
+    uid: 65532
+    gid: 65532
+    permissions: 0o755
+    recursive: true
```
Comment (rawlingsj) [inline images/opensearch/config/template.apko.yaml:46]: I know we shouldn't need these, just adding everything for now and dial back once things are working

File: images/opensearch/config/template.apko.yaml [hunk 2]
```diff
@@ -20,9 +20,36 @@ paths:
...
+
+work-dir: /usr/share/opensearch
 
 environment:
   JAVA_HOME: /usr/lib/jvm/default-jvm
+  OPENSEARCH_JAVA_OPTS: -Xlog:gc:/usr/share/opensearch/logs/gc.log
```
Comment (rawlingsj) [inline images/opensearch/config/template.apko.yaml:52]: I would hope we don't need this and by creating the folders above we can remove this

File: images/opensearch/tests/main.tf [hunk 3]
```diff
@@ -1,14 +1,74 @@
...
+
+      appender.console.type = Console
+      appender.console.name = console
+      appender.console.layout.type = PatternLayout
+      appender.console.layout.pattern = [%d{ISO8601}][%-5p][%-25c{1.}] [%node_name]%marker %m%n
+
+      rootLogger.level = info
+      rootLogger.appenderRef.console.ref = console
+      EOT
+    }
```
Comment (rawlingsj) [inline images/opensearch/tests/main.tf:55]: Not sure we need this, trying to fix the JVM gc.log permission issue

**Comment by rawlingsj** [comment]
@pdeslaur @found-it I felt I was getting somewhere then when testing locally the k3s cluster can't pull the image even when it's built and pullable from my host.  Might be something up with my config or something I've done in this PR.

If we can get the opensearch image tests working on Kubernetes then we can test logstash and opensearch dashboard integrations.

**Comment by rawlingsj** [comment]
The error I'm seeing with the opensearch test on kubernetes is
```
kubectl logs opensearch-cluster-master-0 -n opensearch-dashboards
Defaulted container "opensearch" out of: opensearch, fsgroup-volume (init)
Exception in thread "main" java.lang.RuntimeException: starting java failed with [1]
output:
[0.000s][error][logging] Error opening log file 'logs/gc.log': Permission denied
[0.000s][error][logging] Initialization of output 'file=logs/gc.log' using options 'filecount=32,filesize=64m' failed.
error:
Invalid -Xlog option '-Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m', see error log for details.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
	at org.opensearch.tools.launchers.JvmErgonomics.flagsFinal(JvmErgonomics.java:125)
	at org.opensearch.tools.launchers.JvmErgonomics.finalJvmOptions(JvmErgonomics.java:87)
	at org.opensearch.tools.launchers.JvmErgonomics.choose(JvmErgonomics.java:70)
	at org.opensearch.tools.launchers.JvmOptionsParser.jvmOptions(JvmOptionsParser.java:150)
	at org.opensearch.tools.launchers.JvmOptionsParser.main(JvmOptionsParser.java:108)
```

**Comment by rawlingsj** [comment]
Ok figured out the error pulling image when testing locally, I needed to update my `main.tf` and configure k3s' registry mirror to point to docker's magic host dns...

https://github.com/joshrwolf/images/blob/imagetest-docs/TERRAFORM.md#local-setup

**Comment by rawlingsj** [comment]
So now we have missing plugins

```
[2024-02-09T11:07:33,908][ERROR][o.o.b.OpenSearchUncaughtExceptionHandler] [opensearch-cluster-master-0] uncaught exception in thread [main]
org.opensearch.bootstrap.StartupException: java.lang.IllegalArgumentException: unknown setting [plugins.security.ssl.transport.pemkey_filepath] please check that any required plugins are installed, or check the breaking changes documentation for removed settings
	at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:184) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.bootstrap.OpenSearch.execute(OpenSearch.java:171) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:104) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138) ~[opensearch-cli-2.11.1.jar:2.11.1]
	at org.opensearch.cli.Command.main(Command.java:101) ~[opensearch-cli-2.11.1.jar:2.11.1]
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:137) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:103) ~[opensearch-2.11.1.jar:2.11.1]
Caused by: java.lang.IllegalArgumentException: unknown setting [plugins.security.ssl.transport.pemkey_filepath] please check that any required plugins are installed, or check the breaking changes documentation for removed settings
	at org.opensearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:608) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:549) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:519) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:489) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.common.settings.SettingsModule.<init>(SettingsModule.java:178) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.node.Node.<init>(Node.java:578) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.node.Node.<init>(Node.java:407) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:242) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:242) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:404) ~[opensearch-2.11.1.jar:2.11.1]
	at org.opensearch
```

**Comment by rawlingsj** [comment]
Looks like this is the security plugin https://github.com/opensearch-project/security

**Review by dlorenc — APPROVED** [review]

**Pull request merged**